### ISSUE TITLE
v2.2.0-beta Linux "Failed to write the trace file." on writable folder

### ISSUE THREAD

**Issue by dstj** [microsoft/ApplicationInsights-Profiler-AspNetCore#109]
Hi @xiaomi7732 

As mentioned a bit in #108, I now have Profiler traces uploading to the Azure Portal, but I also got several "Failed to write the trace file" exceptions from the ServiceProfiler.

Here are the two error logs I see (I don't have the Debug logs this time though)

```json
{
	"Timestamp":"2020-10-28T21:46:05.3943433+00:00",
	"Level":"Error",
	"MessageTemplate": "Failed to write the trace file.",
	"RenderedMessage":"Failed to write the trace file.",
	"Exception":"System.IO.IOException: Unable to read data from the transport connection: Operation canceled.
		---> System.Net.Sockets.SocketException (125): Operation canceled
		--- End of inner exception stack trace ---
		at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
		at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
		at System.IO.Stream.CopyToAsyncInternal(Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
		at ServiceProfiler.EventPipe.Client.TraceControls.DiagnosticsClientTraceControl.StartWriteAsync(String traceFilePath, Stream readFrom)",
	"Properties": {
		"SourceContext":"ServiceProfiler.EventPipe.Client.TraceControls.DiagnosticsClientTraceControl",
		"ExceptionDetail":{
			"Type":"System.IO.IOException",
			"HResult":-2146232800,
			"Message":"Unable to read data from the transport connection: Operation canceled.",
			"Source":"System.Net.Sockets",
			"InnerException":{
				"HResult":-2147467259,
				"Message":"Operation canceled",
				"Source":null,
				"SocketErrorCode":"OperationAborted",
				"ErrorCode":125,
				"NativeErrorCode":125,
				"Type":"System.Net.Sockets.SocketException"
			}
		}
	}
}
```

```json
{
	"Timestamp":"2020-10-28T23:53:26.8008666+00:00",
	"Level":"Error",
	"MessageTemplate":"Failed to write the trace file.",
	"RenderedMessage":"Failed to write the trace file.",
	"Exception":"System.ObjectDisposedException: Cannot access a disposed object.
		Object name: 'System.Net.Sockets.NetworkStream'.
		at System.Net.Sockets.NetworkStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
		at System.IO.Stream.CopyToAsyncInternal(Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
		at ServiceProfiler.EventPipe.Client.TraceControls.DiagnosticsClientTraceControl.StartWriteAsync(String traceFilePath, Stream readFrom)",
	"Properties":{
		"SourceContext":"ServiceProfiler.EventPipe.Client.TraceControls.DiagnosticsClientTraceControl",
		"ExceptionDetail":{
			"HResult":-2146232798,
			"Message":"Cannot access a disposed object. Object name: 'System.Net.Sockets.NetworkStream'.",
			"Source":"System.Net.Sockets",
			"ObjectName":"System.Net.Sockets.NetworkStream",
			"Type":"System.ObjectDisposedException"
		}
	}
}
```

It both cases, it originated from `StartWriteAsync()`. The `ObjectDisposedException` is concerning. Also, I know the `/tmp` folder is writable because I do see some traces uploaded to Azure.

Do those traces mean anything to you? Does that help?

Question: Could it be a "file size limit reached" maybe? Is the `BufferSizeInMB` config setting about the tmp file size or is it a memory buffer size?

--
Environment:
Linux (from the Heroku-16 stack)
.NET Core 3.1.403
Microsoft.ApplicationInsights.AspNetCore 2.15.0
Microsoft.ApplicationInsights.WorkerService 2.15.0
Microsoft.ApplicationInsights.Profiler.AspNetCore 2.2.0-beta2

**Comment by xiaomi7732** [comment]
@dstj, thank you again for your feedback.
This actually goes in to the steam of .NET Core Runtime for the Eventpipe. @sywhang could you please help take a look and see if there's anything we can infer from the exception info? I have sent you the caller's code offline.

**Comment by xiaomi7732** [comment]
@dstj I got feedback from @sywhang quickly. We examined the implementation and think there could be 2 potential issues:

1. The application is shutting down upon when the profiler run into stopping.
1. Somehow, the mounted point to /tmp/ is lost - that is less likely.

Could you please comment on the first possibility?

Either way, when the time the exception is thrown, the trace stream is disposed/closed already. I'll update the profiler to avoid throwing the exception in that specific case.

This issue will be used to track the progress for the fix.

**Comment by dstj** [comment]
@xiaomi7732 Thanks for the quick reply.

1. I looked at the timing of the error logs and the server reboots and no, the server **was not** rebooting at that time.

2. That is highly unlikely. I'm on Heroku and they use their [Ephemeral filesystem](https://devcenter.heroku.com/articles/dynos#ephemeral-filesystem), but that basically means that the filesystem works like normal, but will be lost on server reboots, so don't store anything permanent.

> Either way, when the time the exception is thrown, the trace stream is disposed/closed already. I'll update the profiler to avoid throwing the exception in that specific case.

Maybe still log a warning though... ?

**Comment by xiaomi7732** [comment]
> Maybe still log a warning though... ?

Sure.

**Comment by xiaomi7732** [comment]
The fix is in the code base. Will be delivered in the next release.

**Comment by xiaomi7732** [comment]
This is released in [2.2.0-beta3](https://www.nuget.org/packages/Microsoft.ApplicationInsights.Profiler.AspNetCore/2.2.0-beta3).
